zebrad/src/commands/start/sync.rs

```diff
@@ -123,21 +130,41 @@ where
                             break;
                         }
                     }
+                    tracing::debug!(
+                        first_unknown,
+                        "found index of first unknown hash in response"
+                    );
                     if first_unknown == hashes.len() {
+                        // XXX until we fix the TODO above to construct the locator correctly,
+                        // we might hit this case, but it will be unexpected afterwards.
                         tracing::debug!("no new hashes, even though we gave our tip?");
                         continue;
                     }
-                    let unknown_hashes = &hashes[first_unknown..];
-                    download_set.extend(unknown_hashes);
 
-                    // ObtainTips Step 4
-                    //
-                    // Combine the last elements of each list into a set; this
-                    // is the set of prospective tips.
+                    let unknown_hashes = &hashes[first_unknown..];
                     let new_tip = *unknown_hashes
                         .last()
                         .expect("already checked first_unknown < hashes.len()");
-                    let _ = self.prospective_tips.insert(new_tip);
+
+                    // ObtainTips Step 4:
+                    // Combine the last elements of each list into a set; this is the
+                    // set of prospective tips.
+                    if !download_set.contains(&new_tip) {
+                        tracing::debug!(?new_tip, "adding new prospective tip");
+                        self.prospective_tips.insert(new_tip);
+                    } else {
+                        tracing::debug!(?new_tip, "discarding tip already queued for download");
+                    }
```

Comment:
This prevents adding a sub-block of an existing prospective tip as a new prospective tip on its own, similarly to what we do in extendtips.